File: pascal/ipcalc2.pas
program ipcalc2;

{$mode objfpc}{$H+}

uses
  SysUtils;

function ParseIPv4(const S: string): QWord;
var
  I, Octet, Count: Integer;
  Value: QWord;
  Ch: Char;
begin
  Value := 0;
  Octet := 0;
  Count := 0;

  for I := 1 to Length(S) + 1 do
  begin
    if I <= Length(S) then
      Ch := S[I]
    else
      Ch := '.';

    if Ch = '.' then
    begin
      if (Octet < 0) or (Octet > 255) or (Count >= 4) then
        raise Exception.Create('invalid IPv4 address');
      Value := (Value shl 8) or QWord(Octet);
      Inc(Count);
      Octet := 0;
    end
    else if (Ch >= '0') and (Ch <= '9') then
      Octet := Octet * 10 + (Ord(Ch) - Ord('0'))
    else
      raise Exception.Create('invalid IPv4 address');
  end;

  if Count <> 4 then
    raise Exception.Create('invalid IPv4 address');

  Result := Value;
end;

function IPv4ToString(V: QWord): string;
begin
  Result := Format('%d.%d.%d.%d', [
    (V shr 24) and 255,
    (V shr 16) and 255,
    (V shr 8) and 255,
    V and 255
  ]);
end;

var
  IP, Mask, NetworkAddr, BroadcastAddr, FirstAddr, LastAddr, Hosts: QWord;
  Prefix: Integer;
begin
  if ParamCount <> 2 then
  begin
    WriteLn(StdErr, 'usage: ', ParamStr(0), ' IP PREFIX');
    Halt(2);
  end;

  try
    IP := ParseIPv4(ParamStr(1));
    Prefix := StrToInt(ParamStr(2));
  except
    on E: Exception do
    begin
      WriteLn(StdErr, E.Message);
      Halt(2);
    end;
  end;

  if (Prefix < 1) or (Prefix > 30) then
  begin
    WriteLn(StdErr, 'prefix must be between 1 and 30');
    Halt(2);
  end;

  { Parsing is finished. Fill in only these calculations. }
  Mask := 0;          { TODO }
  NetworkAddr := 0;   { TODO }
  BroadcastAddr := 0; { TODO }
  FirstAddr := 0;     { TODO }
  LastAddr := 0;      { TODO }
  Hosts := 0;         { TODO }

  WriteLn('IP: ', IPv4ToString(IP));
  WriteLn('Netmask: ', IPv4ToString(Mask));
  WriteLn('Network: ', IPv4ToString(NetworkAddr));
  WriteLn('Broadcast: ', IPv4ToString(BroadcastAddr));
  WriteLn('First usable: ', IPv4ToString(FirstAddr));
  WriteLn('Last usable: ', IPv4ToString(LastAddr));
  WriteLn('Usable hosts: ', Hosts);
end.

File: c/ipcalc2.c
#include <stdint.h>
#include <stdio.h>
#include <stdlib.h>

static int parse_ipv4(const char *s, uint32_t *out)
{
    unsigned a, b, c, d;
    char tail;

    if (sscanf(s, "%u.%u.%u.%u%c", &a, &b, &c, &d, &tail) != 4)
        return 0;
    if (a > 255 || b > 255 || c > 255 || d > 255)
        return 0;

    *out = ((uint32_t)a << 24) |
           ((uint32_t)b << 16) |
           ((uint32_t)c << 8)  |
           (uint32_t)d;
    return 1;
}

static void print_ipv4(uint32_t v)
{
    printf("%u.%u.%u.%u",
           (v >> 24) & 255,
           (v >> 16) & 255,
           (v >> 8) & 255,
           v & 255);
}

int main(int argc, char **argv)
{
    uint32_t ip;
    unsigned prefix;

    if (argc != 3) {
        fprintf(stderr, "usage: %s IP PREFIX\n", argv[0]);
        return 2;
    }

    if (!parse_ipv4(argv[1], &ip)) {
        fprintf(stderr, "invalid IPv4 address\n");
        return 2;
    }

    prefix = (unsigned)strtoul(argv[2], NULL, 10);
    if (prefix < 1 || prefix > 30) {
        fprintf(stderr, "prefix must be between 1 and 30\n");
        return 2;
    }

    /* Command-line parsing is finished.  The useful inputs are now:
       ip      - the IPv4 address as one 32-bit integer
       prefix  - the CIDR prefix length

       Fill in only the calculations below. */

    uint32_t mask      = 0;  /* TODO */
    uint32_t network   = 0;  /* TODO */
    uint32_t broadcast = 0;  /* TODO */
    uint32_t first     = 0;  /* TODO */
    uint32_t last      = 0;  /* TODO */
    uint32_t hosts     = 0;  /* TODO */

    printf("IP: "); print_ipv4(ip); putchar('\n');
    printf("Netmask: "); print_ipv4(mask); putchar('\n');
    printf("Network: "); print_ipv4(network); putchar('\n');
    printf("Broadcast: "); print_ipv4(broadcast); putchar('\n');
    printf("First usable: "); print_ipv4(first); putchar('\n');
    printf("Last usable: "); print_ipv4(last); putchar('\n');
    printf("Usable hosts: %u\n", hosts);
    return 0;
}
